File: js/src/prompt-cache/prompt-cache.ts
import { Prompt } from "../logger";
import { LRUCache } from "../lru-cache";
import { DiskCache } from "./disk-cache";

/**
 * Identifies a prompt in the cache using either project ID or project name along with the slug.
 */
export interface PromptKey {
  /**
   * The slug identifier for the prompt within its project.
   */
  slug?: string;

  /**
   * The version of the prompt.
   */
  version?: string;

  /**
   * The ID of the project containing the prompt.
   * Either projectId or projectName must be provided.
   */
  projectId?: string;

  /**
   * The name of the project containing the prompt.
   * Either projectId or projectName must be provided.
   */
  projectName?: string;

  /**
   * The ID of a specific prompt. If provided, slug and project parameters are ignored.
   */
  id?: string;
}

export type PromptMemoryCacheEntry = {
  value: Prompt;
  resolvedOrgIdentity?: string;
};

export type PromptDiskCacheEntry = {
  value: ReturnType<Prompt["_internalSerializeForCache"]>;
  resolvedOrgIdentity?: string;
};

/**
 * Creates a unique cache key from prompt key.
 * @param key - The prompt key to convert into a cache key.
 * @returns A string that uniquely identifies the prompt in the cache.
 * @throws {Error} If neither projectId nor projectName is provided (when not using id).
 */
function createCacheKey(key: PromptKey, namespace?: string): string {
  let cacheKey: string;
  if (key.id) {
    // When caching by ID, we don't need project or slug
    cacheKey = `id:${key.id}`;
  } else {
    const prefix = key.projectId ?? key.projectName;
    if (!prefix) {
      throw new Error("Either projectId or projectName must be provided");
    }
    if (!key.slug) {
      throw new Error("Slug must be provided when not using ID");
    }
    cacheKey = `${prefix}:${key.slug}:${key.version ?? "latest"}`;
  }
  return namespace === undefined
    ? cacheKey
    : `${namespace.length}:${namespace}:${cacheKey}`;
}

/**
 * A configurable cache for Braintrust prompts with optional in-memory and filesystem storage.
 *
 * This cache can use either layer independently, both layers together, or no layers.
 */
export class PromptCache {
  private readonly memoryCache?: LRUCache<string, PromptMemoryCacheEntry>;
  private readonly diskCache?: DiskCache<PromptDiskCacheEntry>;
  private readonly namespace?: string;
  private readonly expectedResolvedOrgIdentity?: string;

  constructor(options: {
    memoryCache?: LRUCache<string, PromptMemoryCacheEntry>;
    diskCache?: DiskCache<PromptDiskCacheEntry>;
    namespace?: string;
    expectedResolvedOrgIdentity?: string;
  }) {
    this.memoryCache = options.memoryCache;
    this.diskCache = options.diskCache;
    this.namespace = options.namespace;
    this.expectedResolvedOrgIdentity = options.expectedResolvedOrgIdentity;
  }

  /**
   * Returns a cache view that shares the same storage layers but isolates all
   * entries under the provided namespace.
   */
  withNamespace(
    namespace: string,
    expectedResolvedOrgIdentity?: string,
  ): PromptCache {
    return new PromptCache({
      memoryCache: this.memoryCache,
      diskCache: this.diskCache,
      namespace,
      expectedResolvedOrgIdentity,
    });
  }

  /**
   * Retrieves a prompt from the cache.
   * First checks the in-memory LRU cache, then falls back to checking the disk cache if available.
   */
  async get(key: PromptKey): Promise<Prompt | undefined> {
    const cacheKey = createCacheKey(key, this.namespace);
    if (this.memoryCache) {
      const memoryEntry = this.memoryCache.get(cacheKey);
      if (
        memoryEntry !== undefined &&
        (this.expectedResolvedOrgIdentity === undefined ||
          memoryEntry.resolvedOrgIdentity === this.expectedResolvedOrgIdentity)
      ) {
        return memoryEntry.value;
      }
    }

    if (this.diskCache) {
      const diskEntry = await this.diskCache.get(cacheKey);
      if (
        !diskEntry ||
        (this.expectedResolvedOrgIdentity !== undefined &&
          diskEntry.resolvedOrgIdentity !== this.expectedResolvedOrgIdentity)
      ) {
        return undefined;
      }
      const serializedPrompt = diskEntry.value;
      const diskPrompt = new Prompt(
        serializedPrompt.metadata,
        serializedPrompt.defaults,
        serializedPrompt.noTrace,
      );
      this.memoryCache?.set(cacheKey, {
        value: diskPrompt,
        resolvedOrgIdentity: diskEntry.resolvedOrgIdentity,
      });
      return diskPrompt;
    }

    return undefined;
  }

  /**
   * Stores a prompt in the cache.
   * Writes to the in-memory cache and the disk cache if available.
   *
   * @param key - The key to store the value under.
   * @param value - The value to store in the cache.
   * @throws If there is an error writing to the disk cache.
   */
  async set(key: PromptKey, value: Prompt): Promise<void> {
    const cacheKey = createCacheKey(key, this.namespace);
    const memoryEntry = {
      value,
      resolvedOrgIdentity: this.expectedResolvedOrgIdentity,
    };
    this.memoryCache?.set(cacheKey, memoryEntry);
    if (this.diskCache) {
      await this.diskCache.set(cacheKey, {
        value: value._internalSerializeForCache(),
        resolvedOrgIdentity: this.expectedResolvedOrgIdentity,
      });
    }
  }
}

File: py/src/braintrust/prompt_cache/prompt_cache.py
"""
This module implements a two-layer caching system for Braintrust prompts.

The caching system consists of:
1. A fast in-memory LRU cache for frequently accessed prompts
2. A persistent disk-based cache that serves as a backing store

This allows for efficient prompt retrieval while maintaining persistence across sessions.
The cache is keyed by project identifier (ID or name), prompt slug, and version.
"""


from braintrust import prompt
from braintrust.prompt_cache import disk_cache, lru_cache


def _create_cache_key(
    project_id: str | None,
    project_name: str | None,
    slug: str | None,
    version: str = "latest",
    id: str | None = None,
) -> str:
    """Creates a unique cache key from project identifier, slug and version, or from ID."""
    if id:
        # When caching by ID, we don't need project or slug
        return f"id:{id}"

    prefix = project_id or project_name
    if not prefix:
        raise ValueError("Either project_id or project_name must be provided")
    if not slug:
        raise ValueError("Slug must be provided when not using ID")
    return f"{prefix}:{slug}:{version}"


class PromptCache:
    """
    A two-layer cache for Braintrust prompts with both in-memory and filesystem storage.

    This cache implements either a one or two-layer caching strategy:
    1. A fast in-memory LRU cache for frequently accessed prompts.
    2. An optional persistent filesystem-based cache that serves as a backing store.
    """

    def __init__(
        self,
        memory_cache: lru_cache.LRUCache[str, prompt.PromptSchema],
        disk_cache: disk_cache.DiskCache[prompt.PromptSchema] | None = None,
    ):
        """
        Initialize the prompt cache.

        Args:
            memory_cache: The memory cache to use.
            disk_cache: Optional disk cache to use as backing store.
        """
        self.memory_cache = memory_cache
        self.disk_cache = disk_cache

    def get(
        self,
        slug: str | None = None,
        version: str = "latest",
        project_id: str | None = None,
        project_name: str | None = None,
        id: str | None = None,
    ) -> prompt.PromptSchema:
        """
        Retrieve a prompt from the cache.

        Args:
            slug: The unique identifier for the prompt within its project. Required if id is not provided.
            version: The version of the prompt. Defaults to "latest".
            project_id: The ID of the project containing the prompt.
            project_name: The name of the project containing the prompt.
            id: The ID of a specific prompt. If provided, slug and project parameters are ignored.

        Returns:
            The cached Prompt object.

        Raises:
            ValueError: If neither project_id nor project_name is provided (when not using id).
            KeyError: If the prompt is not found in the cache.
        """
        cache_key = _create_cache_key(project_id, project_name, slug, version, id)

        # First check memory cache.
        try:
            return self.memory_cache.get(cache_key)
        except KeyError:
            pass

        # If not in memory and disk cache exists, check disk cache.
        if self.disk_cache:
            prompt = self.disk_cache.get(cache_key)
            if prompt is None:
                raise KeyError(f"Prompt not found in cache: {cache_key}")

            # Store in memory cache.
            self.memory_cache.set(cache_key, prompt)
            return prompt

        raise KeyError(f"Prompt not found in cache: {cache_key}")

    def set(
        self,
        value: prompt.PromptSchema,
        slug: str | None = None,
        version: str = "latest",
        project_id: str | None = None,
        project_name: str | None = None,
        id: str | None = None,
    ) -> None:
        """
        Store a prompt in the cache.

        Args:
            slug: The unique identifier for the prompt within its project. Required if id is not provided.
            version: The version of the prompt. Defaults to "latest".
            value: The Prompt object to store.
            project_id: The ID of the project containing the prompt.
            project_name: The name of the project containing the prompt.
            id: The ID of a specific prompt. If provided, slug and project parameters are ignored.

        Raises:
            ValueError: If neither project_id nor project_name is provided (when not using id).
            RuntimeError: If there is an error writing to the disk cache.
        """
        cache_key = _create_cache_key(project_id, project_name, slug, version, id)

        # Update memory cache.
        self.memory_cache.set(cache_key, value)

        # Update disk cache if available.
        if self.disk_cache:
            self.disk_cache.set(cache_key, value)
